compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -2842,6 +2842,34 @@ impl<'tcx> Ty<'tcx> {
         }
     }
 
+    pub fn is_trivially_non_zst(self, tcx: TyCtxt<'tcx>) -> bool {
+        match self.kind() {
+            ty::Infer(ty::IntVar(_) | ty::FloatVar(_))
+            | ty::Uint(_)
+            | ty::Int(_)
+            | ty::Bool
+            | ty::Float(_)
+            | ty::FnPtr(_)
+            | ty::RawPtr(..)
+            | ty::Char
+            | ty::Ref(..)
+            | ty::Closure(..) => true,
+            ty::Array(ty, len) => {
+                if let Some(len) = len.try_to_target_usize(tcx) {
+                    return len > 0 && ty.is_trivially_non_zst(tcx);
+                }
+                false
+            }
+            ty::Tuple(tys) => tys.iter().any(|ty| ty.is_trivially_non_zst(tcx)),
+            ty::Adt(def, args) => def.all_fields().any(|field| {
+                let ty = field.ty(tcx, args);
+                ty.is_trivially_non_zst(tcx)
```

Comment:
This is non-obvious and deserves a comment. It relies on `Result<(), (!, i32)>` not removing the `Err` variant entirely (which plausibly it could do due to it being uninhabited -- but that would also cause issues in MIR).